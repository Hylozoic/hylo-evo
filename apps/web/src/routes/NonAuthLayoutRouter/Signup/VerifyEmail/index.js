import component from './VerifyEmail'

export default component
