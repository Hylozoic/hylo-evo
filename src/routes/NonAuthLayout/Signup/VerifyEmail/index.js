import component from './VerifyEmail'
import connector from './VerifyEmail.connector'
export default connector(component)
