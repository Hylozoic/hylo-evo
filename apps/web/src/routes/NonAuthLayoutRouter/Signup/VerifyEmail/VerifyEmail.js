import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link, Navigate } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import ReactCodeInput from 'react-code-input'
import { formatError } from '../../util'
import getMe from 'store/selectors/getMe'
import getQuerystringParam from 'store/selectors/getQuerystringParam'
import { sendEmailVerification as sendEmailVerificationAction, verifyEmail } from '../Signup.store'
import Loading from 'components/Loading'
import classes from '../Signup.module.scss'

export default function VerifyEmail (props) {
  const dispatch = useDispatch()
  const currentUser = useSelector(getMe)
  const email = currentUser?.email || getQuerystringParam('email', props)
  const token = getQuerystringParam('token', props)
  const [error, setError] = useState()
  const [code, setCode] = useState('')
  const [redirectTo, setRedirectTo] = useState()
  const [loading, setLoading] = useState(false)
  const { t } = useTranslation()

  const sendEmailVerification = async email => {
    const { payload } = await dispatch(sendEmailVerificationAction(email))
    const { success, error } = payload.getData()

    if (error) setError(error)

    if (success) setRedirectTo(`/signup/verify-email?email=${encodeURIComponent(email)}`)
  }

  useEffect(() => {
    if (token) submit()
  }, [])

  if (!email) return <Navigate to='/signup' replace />

  const submit = async value => {
    try {
      setLoading(true)
      const result = await dispatch(verifyEmail(email, value || code, token))
      const error = result?.payload?.getData()?.error

      if (error) setError(error)
    } catch (requestError) {
      // Resolver errors are caught and sent-on as `payload.verifyEmail.error`
      // so this `catch` is for the the case of a network availability or a server
      // implementation issue. It probably can and maybe should be removed.
      setRedirectTo(`/signup?error=${requestError.message}`)
    } finally {
      setLoading(false)
    }
  }

  const handleChange = (value) => {
    if (value.length === 6) {
      submit(value)
    }
    setCode(value)
  }

  if (redirectTo) return <Navigate to={redirectTo} replace />

  if (loading) return <Loading />

  return (
    <div className={classes.form}>
      <Link to='/signup' className={classes.backButton}>&#8592; {t('back')}</Link>
      <div className={classes.formWrapper}>
        <h1 className={classes.title}>{t('Check your email')}</h1>
        <p className={classes.sub-header}>{t("We've sent a 6 digit code to {{email}}. The code will expire shortly, so please enter it here soon.", { email })}</p>
        {error && formatError(error, 'Signup', t)}
        <div className={classes.codeWrapper}>
          <ReactCodeInput type='text' fields={6} onChange={handleChange} />
        </div>
      </div>
      <div onClick={() => sendEmailVerification(email)} className={classes.resend}>{t('Resend code')}</div>
    </div>
  )
}
