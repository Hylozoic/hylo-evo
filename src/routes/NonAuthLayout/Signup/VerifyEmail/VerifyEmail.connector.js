import { push } from 'connected-react-router'
import { connect } from 'react-redux'

import getLoginError from 'store/selectors/getLoginError'
import getQuerystringParam from 'store/selectors/getQuerystringParam'
import { verifyEmail } from '../Signup.store'

export function mapStateToProps (state, props) {
  const email = getQuerystringParam('email', state, props)
  return {
    error: getLoginError(state),
    email
  }
}

export const mapDispatchToProps = {
  push,
  verifyEmail
}

export function mergeProps (stateProps, dispatchProps, ownProps) {
  return {
    ...stateProps,
    ...dispatchProps,
    ...ownProps,
    verifyEmail: (email, code) => {
      dispatchProps.verifyEmail(email, code).then(
        () => { dispatchProps.push('/signup/finish') },
        (e) => { /* error is added to the state by login reducer but we need to catch it here too */ }
      )
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)
