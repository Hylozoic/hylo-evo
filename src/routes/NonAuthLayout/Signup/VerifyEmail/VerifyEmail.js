import React, { useState } from 'react'
import ReactCodeInput from 'react-code-input'
import { Link } from 'react-router-dom'

import { formatError } from '../../util'

import '../Signup.scss'

export default function VerifyEmail (props) {
  const { className, email, error } = props
  const [code, setCode] = useState('')

  const submit = (value) => {
    props.verifyEmail(email, value || code)
  }

  const onChange = (value) => {
    if (value.length === 6) {
      submit(value)
    }
    setCode(value)
  }

  return <div className={className}>
    <Link to='/signup' styleName='back-button'>&#8592; back</Link>
    <div styleName='formWrapper'>
      <h1 styleName='title'>Check your email</h1>
      <p styleName='sub-header'>We've sent a 6 digit code to {email}. The code will expire shortly, so please enter it here soon.</p>
      {error && formatError(error, 'Signup')}
      <div styleName='codeWrapper'>
        <ReactCodeInput type='text' fields={6} onChange={onChange} />
      </div>
    </div>
  </div>
}
