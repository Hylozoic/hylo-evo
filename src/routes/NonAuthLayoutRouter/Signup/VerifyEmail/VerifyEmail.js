import React, { useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Link, Redirect } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'
import ReactCodeInput from 'react-code-input'
import { formatError } from '../../util'
import getMe from 'store/selectors/getMe'
import getQuerystringParam from 'store/selectors/getQuerystringParam'
import { sendEmailVerification as sendEmailVerificationAction, verifyEmail } from '../Signup.store'
import Loading from 'components/Loading'
import '../Signup.scss'

export default function VerifyEmail (props) {
  const dispatch = useDispatch()
  const currentUser = useSelector(getMe)
  const email = currentUser?.email || getQuerystringParam('email', null, props)
  const token = getQuerystringParam('token', null, props)
  const [error, setError] = useState()
  const [code, setCode] = useState('')
  const [redirectTo, setRedirectTo] = useState()
  const [loading, setLoading] = useState(false)
  const { t } = useTranslation()

  const sendEmailVerification = async email => {
    const { payload } = await dispatch(sendEmailVerificationAction(email))
    const { success, error } = payload.getData()

    if (error) setError(error)

    if (success) setRedirectTo(`/signup/verify-email?email=${encodeURIComponent(email)}`)
  }

  useEffect(() => {
    if (token) submit()
  }, [])

  if (!email) return <Redirect to='/signup' />

  const submit = async value => {
    try {
      setLoading(true)
      const result = await dispatch(verifyEmail(email, value || code, token))
      const error = result?.payload?.getData()?.error

      if (error) setError(error)
    } catch (requestError) {
      // Resolver errors are caught and sent-on as `payload.verifyEmail.error`
      // so this `catch` is for the the case of a network availability or a server
      // implementation issue. It probably can and maybe should be removed.
      setRedirectTo(`/signup?error=${requestError.message}`)
    } finally {
      setLoading(false)
    }
  }

  const handleChange = (value) => {
    if (value.length === 6) {
      submit(value)
    }
    setCode(value)
  }

  if (redirectTo) return <Redirect to={redirectTo} />

  if (loading) return <Loading />

  return (
    <div styleName='form'>
      <Link to='/signup' styleName='back-button'>&#8592; {t('back')}</Link>
      <div styleName='formWrapper'>
        <h1 styleName='title'>{t('Check your email')}</h1>
        <p styleName='sub-header'>{t("We've sent a 6 digit code to {{email}}. The code will expire shortly, so please enter it here soon.", { email })}</p>
        {error && formatError(error, 'Signup', t)}
        <div styleName='codeWrapper'>
          <ReactCodeInput type='text' fields={6} onChange={handleChange} />
        </div>
      </div>
      <div onClick={() => sendEmailVerification(email)} styleName='resend'>{t('Resend code')}</div>
    </div>
  )
}
