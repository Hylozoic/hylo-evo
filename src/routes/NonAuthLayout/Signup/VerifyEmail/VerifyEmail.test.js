import React from 'react'
import { shallow } from 'enzyme'
import VerifyEmail from './VerifyEmail'

describe('VerifyEmail', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<VerifyEmail email='test@wheee.com' />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly with an error', () => {
    const wrapper = shallow(<VerifyEmail email='test@wheee.com' error='some error' />)
    expect(wrapper).toMatchSnapshot()
  })
})
