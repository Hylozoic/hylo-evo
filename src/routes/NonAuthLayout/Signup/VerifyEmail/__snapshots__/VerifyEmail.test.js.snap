// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmail renders correctly 1`] = `
<div>
  <Link
    data-stylename="back-button"
    to="/signup"
  >
    ← back
  </Link>
  <div
    data-stylename="formWrapper"
  >
    <h1
      data-stylename="title"
    >
      Check your email
    </h1>
    <p
      data-stylename="sub-header"
    >
      We've sent a 6 digit code to 
      test@wheee.com
      . The code will expire shortly, so please enter it here soon.
    </p>
    <div
      data-stylename="codeWrapper"
    >
      <t
        autoFocus={true}
        disabled={false}
        fields={6}
        filterChars={
          Array [
            "-",
            ".",
          ]
        }
        filterKeyCodes={
          Array [
            189,
            190,
          ]
        }
        forceUppercase={false}
        isValid={true}
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`VerifyEmail renders correctly with an error 1`] = `
<div>
  <Link
    data-stylename="back-button"
    to="/signup"
  >
    ← back
  </Link>
  <div
    data-stylename="formWrapper"
  >
    <h1
      data-stylename="title"
    >
      Check your email
    </h1>
    <p
      data-stylename="sub-header"
    >
      We've sent a 6 digit code to 
      test@wheee.com
      . The code will expire shortly, so please enter it here soon.
    </p>
    <div
      data-stylename="error"
    >
      some error
    </div>
    <div
      data-stylename="codeWrapper"
    >
      <t
        autoFocus={true}
        disabled={false}
        fields={6}
        filterChars={
          Array [
            "-",
            ".",
          ]
        }
        filterKeyCodes={
          Array [
            189,
            190,
          ]
        }
        forceUppercase={false}
        isValid={true}
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
